File: src/main/java/main/core/procedures/equivalence/Eqv.java
package main.core.procedures.equivalence;

import main.core.ast.SCMBoolean;
import main.core.procedures.AFn;
import main.core.procedures.math.IOperation;

public class Eqv extends AFn implements IOperation {

  @Override
  public SCMBoolean invoke(Object... args) {
    Boolean result = zero();
    if (args != null && args.length > 1) {
      for (int i = 0; i < args.length - 1; i++) {
        result = result && apply(args[i], args[i + 1]);
      }
    }
    return SCMBoolean.toSCMBoolean(result);
  }

  public Boolean zero() {
    return Boolean.TRUE;
  }

  public Boolean apply(Object first, Object second) {

    if (first instanceof Character && second instanceof Character) {
      return ((Character) first).equals((Character)second);
    } else if (first instanceof Number && second instanceof Number) {
      return ((Number) first).equals((Number)second);
    }
    return first == second;
  }

  @Override
  public Object call() throws Exception {
    return invoke();
  }

  @Override
  public void run() {
    invoke();
  }
}

File: src/main/java/core/procedures/equivalence/Eqv.kt
package core.procedures.equivalence

import core.procedures.AFn
import core.procedures.Arity.AtLeast

class Eqv : AFn<Any?, Boolean>(name = "eqv?", isPure = true, arity = AtLeast(2)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) = when {
        arg1 is Collection<*> && arg1.isEmpty() && arg2 is Collection<*> && arg2.isEmpty() -> true
        arg1 is Char && arg2 is Char -> arg1 == arg2
        arg1 is Number && arg2 is Number -> arg1 == arg2
        else -> arg1 === arg2
    }

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        2    -> invoke(args[0], args[1])
        else -> (0..args.size - 2).all { invoke(args[it], args[it + 1]) }
    }
}
